pkg/commands/plan/plan_run.go

```diff
@@ -337,56 +380,67 @@ func (c *PlanRunCommand) handleTerraformPlan(ctx context.Context) (*RunResult, e
 		return &RunResult{commentDetails: stderr.String()}, fmt.Errorf("failed to plan: %w", err)
 	}
 
-	c.actions.Group("Formatting plan output")
-
-	_, err = c.terraformClient.Show(ctx, multiStdout, multiStderr, &terraform.ShowOptions{File: util.Ptr(c.planFilename), NoColor: util.Ptr(true)})
-	if err != nil {
+	if _, err = c.withOutGroup("Formatting plan output",
+		func() (int, error) {
+			return c.terraformClient.Show(ctx, multiStdout, multiStderr, &terraform.ShowOptions{File: util.Ptr(c.planFilename), NoColor: util.Ptr(true)}) //nolint:wrapcheck // Want passthrough
+		},
+	); err != nil {
 		return &RunResult{
 			commentDetails: stderr.String(),
 			hasChanges:     hasChanges,
 		}, fmt.Errorf("failed to terraform show: %w", err)
 	}
 
 	stderr.Reset()
-	c.actions.EndGroup()
 
 	githubOutput := terraform.FormatOutputForGitHubDiff(stdout.String())
 
 	planFilePath := path.Join(c.planChildPath, c.planFilename)
 
 	planData, err := os.ReadFile(planFilePath)
 	if err != nil {
-		return &RunResult{
-				hasChanges: hasChanges,
-			},
-			fmt.Errorf("failed to read plan binary: %w", err)
+		return &RunResult{hasChanges: hasChanges}, fmt.Errorf("failed to read plan binary: %w", err)
 	}
 
-	c.actions.Group("Uploading plan file")
+	bucketObjectPath := fmt.Sprintf("guardian-plans/%s/%s/%d/%s", c.GitHubFlags.FlagGitHubOwner, c.GitHubFlags.FlagGitHubRepo, c.flagPullRequestNumber, planFilePath)
+	c.Outf("Uploading plan file to gs://%s/%s", c.flagBucketName, bucketObjectPath)
 
-	if err := c.handleUploadGuardianPlan(ctx, planFilePath, planData, planExitCode); err != nil {
+	if err := c.handleUploadGuardianPlan(ctx, bucketObjectPath, planData, planExitCode); err != nil {
 		return &RunResult{hasChanges: hasChanges}, fmt.Errorf("failed to upload plan data: %w", err)
 	}
 
 	stderr.Reset()
-	c.actions.EndGroup()
 
 	return &RunResult{
 		commentDetails: githubOutput,
 		hasChanges:     hasChanges,
 	}, nil
 }
 
-// handleUploadGuardianPlan uploads the Guardian plan binary to the configured Guardian storage bucket.
-func (c *PlanRunCommand) handleUploadGuardianPlan(ctx context.Context, planFilePath string, planData []byte, planExitCode int) error {
-	guardianPlanName := fmt.Sprintf("guardian-plans/%s/%s/%d/%s", c.GitHubFlags.FlagGitHubOwner, c.GitHubFlags.FlagGitHubRepo, c.flagPullRequestNumber, planFilePath)
+// withOutGroup runs a function and ensures it is wrapped in GitHub actions
+// grouping syntax. If this is not in an action, output is printed without grouping syntax.
+func (c *PlanRunCommand) withOutGroup(msg string, fn outGroupFunc) (int, error) {
+	if c.GitHubFlags.FlagIsGitHubActions {
+		c.actions.Group(msg)
+	} else {
+		c.Outf(msg)
+	}
 
-	metadata := make(map[string]string)
-	metadata["plan_exit_code"] = strconv.Itoa(planExitCode)
+	exitCode, err := fn()
+
+	if c.GitHubFlags.FlagIsGitHubActions {
```

Comment:
Should this be wrapped in a deferred function instead? Im not sure it matters in this case.

Reply:
I would defer; otherwise you run the risk of `fn` panicing and then never closing the group. It's also just more readable (I think). So I'd do:

```go
if c.GitHubFlags.FlagIsGitHubActions {
  c.actions.Group(msg)
  defer c.actions.EndGroup()
} else {
  c.Outf(msg)
}
```